### PR TITLE
lstm: micro speech model with lstm op running on HiFi5 DSP

### PR DESCRIPTION
**Pull request by nyadla-sys**
Ported lstm op from TF GitHub
    BUG= microspeech model with lstm op on HiFi5 DSP

    Signed-off-by: Lukman Rahumathulla <lukmanr@cadence.com>
    Signed-off-by: Manjunath CP <manjup@cadence.com>
    Signed-off-by: Niranjan Yadla <nyadla@cadence.com>
    Signed-off-by: Raj Pawate <pawateb@cadence.com>

### CODE REVIEW

**Comment by nyadla-sys** [comment]
@petewarden:Can you please review and merge this PR into TFLM tree?

**Comment by nyadla-sys** [comment]
ci:run

**Review by petewarden — COMMENTED** [review]
Thanks for this! It's very helpful to see it all laid out. Here are some high-level thoughts:

 - Can you break out the example into a separate PR we could have as a followup?
 - It appears there are some changes to the core library code, as contained in the files_to_be_copied directory. Can you make those changes to the files themselves, so we can see the diffs needed?

**Comment by nyadla-sys** [comment]
@petewarden 
removed example code, will create new PR after this PR gets merged 
Added the code changes to the core library code.

File: tensorflow/lite/kernels/kernel_util.h [hunk 1]
```diff
@@ -119,6 +119,20 @@ TfLiteStatus GetTemporarySafe(const TfLiteContext* context,
...
+//   TfLiteTensor* my_tensor;
+//   TF_LITE_ENSURE_OK(context,
+//                     GetTemporarySafe(context, node, kMyTensorIdx,
+//                     &my_tensor));
+//   // can use my_tensor directly from here onwards, it is not nullptr
+//
+// Should be used in cases where the binary size is too large.
+TfLiteStatus GetScratchSafe(const TfLiteContext* context,
+                            const TfLiteNode* node, int index,
+                            TfLiteTensor** tensor);
```
Comment (petewarden) [inline tensorflow/lite/kernels/kernel_util.h:134]: Is this change required, or just a nice-to-have?

File: tensorflow/lite/kernels/op_macros.h [hunk 2]
```diff
@@ -15,23 +15,69 @@ limitations under the License.
 #ifndef TENSORFLOW_LITE_KERNELS_OP_MACROS_H_
 #define TENSORFLOW_LITE_KERNELS_OP_MACROS_H_
 
+// If we're on a platform without standard IO functions, fall back to a
+// non-portable function.
+#ifdef TF_LITE_MCU_DEBUG_LOG
+
```
Comment (petewarden) [inline tensorflow/lite/kernels/op_macros.h:21]: This change is a bit surprising, can you add some details about why this is needed? Is it because porting the LSTM op brought a requirement for new macros to be supported?

File: tensorflow/lite/micro/kernels/kernel_util.cc [hunk 3]
```diff
@@ -13,13 +13,33 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
-#include "tensorflow/lite/micro/kernels/kernel_util.h"
-
 #include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
```
Comment (petewarden) [inline tensorflow/lite/micro/kernels/kernel_util.cc:17]: We'll probably want to revert the header order back to the original here, since kernel_util.h should be able to compile without common.h being included first, and including it first is a way to check that.

File: tensorflow/lite/micro/kernels/kernel_util.h [hunk 4]
```diff
@@ -18,6 +18,8 @@ limitations under the License.
 
 #include <cstdint>
 
+#define FLATBUFFERS_LOCALE_INDEPENDENT 0
```
Comment (petewarden) [inline tensorflow/lite/micro/kernels/kernel_util.h:21]: It seems a bit odd to be including flexbuffer utility functions in this file, and to have to define a macro in the code before we include the header. We should probably move these functions to their own .h and .cc files. Is there a reason that TFL mobile doesn't already require these?

File: tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.cc [hunk 5]
```diff
@@ -0,0 +1,775 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```
Comment (petewarden) [inline tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.cc:1]: Is this a copy of an existing file in the tree? It's hard to tell what's different without a diff view.

File: tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h [hunk 6]
```diff
@@ -0,0 +1,320 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```
Comment (petewarden) [inline tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h:1]: As above, is this a copied file?

File: tensorflow/lite/micro/micro_allocator.h [hunk 7]
```diff
@@ -240,6 +247,12 @@ class MicroAllocator {
 
   ErrorReporter* error_reporter() const;
 
+  // Returns the first subgraph from the model.
+  const SubGraph* GetSubGraphFromModel(const Model* model);
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_allocator.h:251]: Can you explain the need for this GetSubGraphFromModel function?

File: tensorflow/lite/micro/micro_interpreter.cc [hunk 8]
```diff
@@ -31,8 +32,36 @@ limitations under the License.
 #include "tensorflow/lite/schema/schema_generated.h"
 #include "tensorflow/lite/schema/schema_utils.h"
 
+//#include "stdio.h"
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_interpreter.cc:35]: This seems like something left from debugging?

File: tensorflow/lite/micro/micro_interpreter.cc [hunk 9]
```diff
@@ -31,8 +32,36 @@ limitations under the License.
...
 namespace tflite {
+template <class T>
+std::vector<int> FlatBufferIntArrayToVector(T* flat_array) {
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_interpreter.cc:38]: We try to avoid any use of std::vectors or other STL containers, because they usually bring in a lot of code and require dynamic memory allocation. Can you talk a bit about why this is needed in this case?

File: tensorflow/lite/micro/micro_interpreter.cc [hunk 10]
```diff
@@ -175,6 +206,46 @@ TfLiteStatus MicroInterpreter::PrepareNodeAndRegistrationDataFromFlatbuffer() {
       node->builtin_data = reinterpret_cast<void*>(builtin_data);
       node->custom_initial_data = custom_data;
       node->custom_initial_data_size = custom_data_size;
+
+      // LSTM specific COde
+      size_t alloc_count = subgraph->tensors()->size();
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_interpreter.cc:211]: It might be useful in the design doc to give an overview of what you've found makes the LSTM op different from regular ops, and so what support is required. From the code here it looks like "intermediates" are important, but I'd love to know what you've found about what that means in detail.

File: tensorflow/lite/micro/micro_interpreter.cc [hunk 11]
```diff
@@ -326,6 +397,117 @@ TfLiteStatus MicroInterpreter::RequestScratchBufferInArena(TfLiteContext* ctx,
       bytes, interpreter->graph_.GetCurrentSubgraphIndex(), buffer_idx);
 }
 
+TfLiteStatus MicroInterpreter::ResizeTensor(TfLiteContext* context,
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_interpreter.cc:400]: One of the design decisions we made with TFL Micro was only to support graphs with fixed shapes throughout. Can you add some context about why this is needed in this case?

File: tensorflow/lite/micro/micro_interpreter.h [hunk 12]
```diff
@@ -110,6 +116,10 @@ class MicroInterpreter {
 
   TfLiteStatus initialization_status() const { return initialization_status_; }
 
+  size_t operators_size() const {
+    return model_->subgraphs()->Get(0)->operators()->size();
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_interpreter.h:120]: Can you explain a bit about why the operator size is needed for LSTMs?

File: tensorflow/lite/micro/tools/make/Makefile [hunk 13]
```diff
@@ -147,10 +147,10 @@ COMMON_FLAGS := \
   -ffunction-sections \
   -fdata-sections \
   -fmessage-length=0 \
-  -DTF_LITE_STATIC_MEMORY \
```
Comment (petewarden) [inline tensorflow/lite/micro/tools/make/Makefile:150]: This removal is a bit surprising, can you add details about what functions are needed.

**Review by petewarden — COMMENTED** [review]
I've left individual comments, but some high-level ideas are:
 - Can you write up a public design doc that explains the goals and approaches you're taking?
 - It looks like there are still a lot of files that look like copies in t/l/m/kertnels/xtensa/lstm/helper_utils/. Can you give more details on those, and if possible move them back to their source positions so we can see diffs?
 - Once I understand more about the changes, we'll probably want to split this up into several PRs, but the design doc will help.

Thanks for putting this together, and your patience with the process, I know this is a lot to take on!

File: tensorflow/lite/kernels/kernel_util.cc [hunk 14]
```diff
@@ -498,42 +524,42 @@ TfLiteStatus CalculateShapeForBroadcast(TfLiteContext* context,
 int TfLiteTypeGetSize(TfLiteType type) {
   switch (type) {
     case kTfLiteUInt8:
-      static_assert(sizeof(uint8_t) == 1, "");
```
[the remark this answers is not in the document]
Comment (nyadla-sys) [reply tensorflow/lite/kernels/kernel_util.cc:501]: ci:run throws error saying assert must not be used for TFLM and will change back to static_assert

File: tensorflow/lite/c/common.h [hunk 15]
```diff
@@ -738,6 +738,8 @@ typedef struct TfLiteContext {
   // WARNING: This is an experimental interface that is subject to change.
   void* (*AllocatePersistentBuffer)(struct TfLiteContext* ctx, size_t bytes);
 
+  void* (*AllocateTempBuffer)(struct TfLiteContext* context, size_t bytes);
```
[the remark this answers is not in the document]
Comment (nyadla-sys) [reply tensorflow/lite/c/common.h:741]: Removed AllocateTempBuffer  API as it is not used in our code.
Used GetScratchSafe API and that calls  GetScratchTensor as part of prepare of unidirectional LSTM  to get  scratch buffer, and we are not sure if this can be replaced with RequestScratchBufferInArena

File: tensorflow/lite/kernels/kernel_util.cc [hunk 16]
```diff
@@ -42,12 +43,22 @@ namespace {
 inline TfLiteTensor* GetTensorAtIndex(const TfLiteContext* context,
                                       int tensor_index) {
   if (context->tensors != nullptr) {
-    return &context->tensors[tensor_index];
+    return context->GetTensor(
```
[the remark this answers is not in the document]
Comment (nyadla-sys) [reply tensorflow/lite/kernels/kernel_util.cc:46]: If we use "return &context->tensors[tensor_index];" xtensa run is crashing ,so changed to context->GetTensor(context, tensor_index);  to solve issue.

File: tensorflow/lite/kernels/kernel_util.cc [hunk 17]
```diff
@@ -435,14 +457,16 @@ TfLiteStatus CalculateShapeForBroadcast(TfLiteContext* context,
   const int dims1 = NumDimensions(input1);
   const int dims2 = NumDimensions(input2);
   const int out_dims = std::max(dims1, dims2);
+  tflite::MicroErrorReporter micro_error_reporter;
```
[the remark this answers is not in the document]
Comment (nyadla-sys) [reply tensorflow/lite/kernels/kernel_util.cc:460]: Yes and LSTM prepare api is called to get GetIntermediates.
will remove micro_error_reporter and replace with  context->ReportError

File: tensorflow/lite/kernels/kernel_util.h (hunk 1 above)
Comment (nyadla-sys) [reply tensorflow/lite/kernels/kernel_util.h:134]: this API is required to get scratch buffer at prepare stage of LSTM op

File: tensorflow/lite/kernels/op_macros.h (hunk 2 above)
Comment (nyadla-sys) [reply tensorflow/lite/kernels/op_macros.h:21]: lstm_eval.cc uses TF_LITE_ASSERT MACRO, so this is required

File: tensorflow/lite/micro/kernels/kernel_util.cc (hunk 3 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/kernels/kernel_util.cc:17]: clang-format3.9 produced this output ,anyway i will revert back

File: tensorflow/lite/micro/kernels/kernel_util.h (hunk 4 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/kernels/kernel_util.h:21]: Sorry we did not use any  flexbuffer utility functions,so we will revert this change

File: tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h (hunk 6 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h:1]: lstm uses some functions from this file

File: tensorflow/lite/micro/micro_allocator.h (hunk 7 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_allocator.h:251]: As part of this PrepareNodeAndRegistrationDataFromFlatbuffer() function , we called AllocateIntermediateTfLiteTensorFromFlatbuffer() function to get intermediate tensor and intern this function uses GetSubGraphFromModel to init TfLiteTensorFromFlatbuffer

File: tensorflow/lite/micro/micro_interpreter.cc (hunk 8 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:35]: reverted this change

File: tensorflow/lite/micro/micro_interpreter.cc (hunk 9 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:38]: DO you have any substitute to this function call or its implementation   tflite::FlatBufferIntArrayToVector(op->intermediates())

File: tensorflow/lite/micro/micro_interpreter.cc (hunk 10 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:211]: we try to use similar to TfLite implementation and I also need sometime to understand indepth details

File: tensorflow/lite/micro/micro_interpreter.cc (hunk 11 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:400]: lstm prepare function uses ResizeTensor(), so we implemented this and it is same as TFLite implementation

File: tensorflow/lite/micro/micro_interpreter.h (hunk 12 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.h:120]: its not required and will revert this change

File: tensorflow/lite/micro/tools/make/Makefile (hunk 13 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/tools/make/Makefile:150]: for example GetIntermediatesSafe,TfLiteIntArrayCopy,TfLiteIntArrayCreate,GetScratchSafe thise functions are called from lstm and are defined only if undef of TF_LITE_STATIC_MEMORY flag

**Comment by nyadla-sys** [comment]
@petewarden @njeffrie : Do you know any op which uses intermediate buffer (apart from lstm).
Removed all added API's from micro_allocator
Please review and let me know further comments..

File: tensorflow/lite/micro/micro_interpreter.cc [hunk 18]
```diff
@@ -175,6 +176,28 @@ TfLiteStatus MicroInterpreter::PrepareNodeAndRegistrationDataFromFlatbuffer() {
       node->builtin_data = reinterpret_cast<void*>(builtin_data);
       node->custom_initial_data = custom_data;
       node->custom_initial_data_size = custom_data_size;
+
```
Comment (petewarden) [inline tensorflow/lite/micro/micro_interpreter.cc:179]: Do you still need this code? It looks like all of the 'eval' code uses TFLiteEvalTensors now, so I'm hoping we can delete this since I think it's the main outstanding change that we'd like to avoid in the core framework code.

Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:179]: Prepare() function of unidirectional_sequence_lstm.cc  is still using TFLiteTensor and using intermediates to get quantization scale and zeropoint

Comment (petewarden) [reply tensorflow/lite/micro/micro_interpreter.cc:179]: I'm hoping you should be able to use context->GetTensor() in prepare, which doesn't use context->tensors, but instead allocates a temporary from the arena:
https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/micro_interpreter.cc#L349

Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:179]: @petewarden  
Please review the below micro interpreter code and I assume we need to add op->intermediates() to node->intermediates after converting to TfLiteIntArray

Comment (petewarden) [reply tensorflow/lite/micro/micro_interpreter.cc:179]: Thanks! That seems a lot more minimal, I think the change should be fine now. Do you still need the include line at the top?

```
#include "tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/util.h"
```

Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:179]: As this OpUsesIntermediates() function is implemented at tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/util.h
do you want me to move this implementation to any specific existing header?

Comment (petewarden) [reply tensorflow/lite/micro/micro_interpreter.cc:179]: Can you check whether the length of the op->intermediates array is more than zero, instead of checking for specific ops?

Comment (nyadla-sys) [reply tensorflow/lite/micro/micro_interpreter.cc:178]: @petewarden 
Added op->intermediates array is more than zero , and this and removed header
#include "tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/util.h"

File: tensorflow/lite/micro/tools/make/Makefile (hunk 13 above)
Comment (nyadla-sys) [reply tensorflow/lite/micro/tools/make/Makefile:150]: @petewarden 
added TF_LITE_STATIC_MEMORY  flag as it is in original

Comment (petewarden) [reply tensorflow/lite/micro/tools/make/Makefile:150]: Thanks! Could you move it so that it's after -fmessage-length=0 \ as it was originally? That will help reduce noise in the diff.

File: tensorflow/lite/micro/micro_interpreter.cc (hunk 18 above)
Comment (petewarden) [reply tensorflow/lite/micro/micro_interpreter.cc:178]: Thanks, this part of the change looks good now.

File: tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h (hunk 6 above)
Comment (petewarden) [reply tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h:1]: From the email discussion with @advaitjain, I'm wondering if you still need these copies of TFLite files, or if you can use the originals now? I think this applies to portable_tensor_utils.*, tensor_utils.*, and util.*

Comment (nyadla-sys) [reply tensorflow/lite/micro/kernels/xtensa/lstm/helper_utils/portable_tensor_utils.h:1]: @petewarden 
Completely removed util.cc and made few changes from the original file of portable_tensor_utils.cc to fix compiler warnings and if you are okay I like to keep it in xtensa/lstm/helper_functions/portable_tensor_utils.,tensor_utils. for sometime and later I will replace our file with @advaitjain PR which will bring these files from Tflite impl.

**Comment by petewarden** [comment]
From inspection, I'm happy with the code now. I'm looking at the CI runs, and I see the following issues:

From the check for TFL files https://github.com/tensorflow/tflite-micro/pull/281/checks?check_run_id=3233873092
```
The following files should be modified in the upstream Tensorflow repo:
tensorflow/lite/c/common.h
tensorflow/lite/core/api/flatbuffer_conversions.cc
tensorflow/lite/core/api/flatbuffer_conversions.h
```

From the build https://github.com/tensorflow/tflite-micro/pull/281/checks?check_run_id=3233875076

```
tensorflow/lite/micro/kernels/xtensa/lstm/kernels/lstm_eval.cc:34:10: fatal error: tensorflow/lite/kernels/internal/tensor_utils_common.h: No such file or directory
   34 | #include "tensorflow/lite/kernels/internal/tensor_utils_common.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It looks like you'll need to run clang format on the changes too, since we're seeing style issues: https://github.com/tensorflow/tflite-micro/pull/281/checks?check_run_id=3233874922

Here's a guide on using clang-format in case it's helpful:
https://github.com/tensorflow/tflite-micro/blob/main/CONTRIBUTING.md#development-workflow-notes

**Comment by nyadla-sys** [comment]
@petewarden 
Please approve this PR

**Comment by petewarden** [comment]
It looks like there are still a few issues with the PR, from the automated tests:
- The code style check is failing, clang-format needs to be run.
- The Cortex M build is failing with errors around restrict and type casts: https://github.com/tensorflow/tflite-micro/pull/281/checks?check_run_id=3297639392
- The project generation isn't working, probably due to some missing files from the makefile lists: https://github.com/tensorflow/tflite-micro/pull/281/checks?check_run_id=3297639123

I'll create a fork of your PR and work on fixing these today, since I think that will probably be easiest. Once I have the changes, I'll have you try them out and ensure your tests are still working.

**Comment by advaitjain** [comment]
> I'll create a fork of your PR and work on fixing these today, since I think that will probably be easiest. Once I have the changes, I'll have you try them out and ensure your tests are still working.

Thanks @petewarden!

Please note that forking this PR is not necessary as you can directly push commits to the current PR:
https://github.com/tensorflow/tflite-micro/blob/main/CONTRIBUTING.md#reviewer-notes

**Comment by petewarden** [comment]
@googlebot I consent.

**Review by petewarden — APPROVED** [review]

**Comment by advaitjain** [comment]
@cad-audio @nyadla-sys: Apologies for missing this during the code review. We would like to avoid having a nested directory structure within `kernels/xtensa`

If you could create a PR that moves the code from `kernels/xtensa/lstm/kernels` directory into the `kernels/xtensa`, that would be appreciated.

Basically, the `kernels/xtensa` folder should have the following files:
```
lstm_eval.cc
lstm_eval.h
lstm_shared.h
unidirectional_sequence_lstm.cc
```

and the `kernels/xtensa/lstm` folder should be deleted.

**Comment by nyadla-sys** [comment]
@advaitjain will create PR with the code changes above
